In [1]:
import pandas as pd
import psycopg2
from sqlalchemy import create_engine

#Establishing the connection

In [2]:
conn = psycopg2.connect(
    dbname="neondb",
    user="neondb_owner",
    password="npg_CeS9fJg2azZD",
    host="ep-falling-glitter-a5m0j5gk-pooler.us-east-2.aws.neon.tech",
    port="5432",
    sslmode="require"
)

# Pandas-kompatible Engine (bequem für read_sql)
engine = create_engine(
    "postgresql+psycopg2://neondb_owner:npg_CeS9fJg2azZD@ep-falling-glitter-a5m0j5gk-pooler.us-east-2.aws.neon.tech:5432/neondb"
)

#Checking out the tables contained in the database

In [3]:
query_tables = """
SELECT table_schema, table_name
FROM information_schema.tables
WHERE table_schema = 'nyc_schools'
ORDER BY table_name;
"""
df_tables = pd.read_sql(query_tables, engine)
df_tables


table_schema                           table_name
0   nyc_schools                anastasia_sat_results
1   nyc_schools                  cleaned_sat_results
2   nyc_schools              darel-kigha_sat_results
3   nyc_schools             hakim-murphy_sat_results
4   nyc_schools           heike_reichert_sat_results
5   nyc_schools                high_school_directory
6   nyc_schools            lars_petschke_sat_results
7   nyc_schools  marianna_gokova_cleaned_sat_results
8   nyc_schools                  meifang_sat_results
9   nyc_schools                              mytable
10  nyc_schools                   sat_scores_cleaned
11  nyc_schools                  school_demographics
12  nyc_schools                 school_safety_report
13  nyc_schools        serhii_sotnichenko_sat_scores
14  nyc_schools             thofa_tazkia_sat_results
15  nyc_schools                  vanessa_sat_results
16  nyc_schools            yana_yelnikova_sat_scores

In [4]:
pd.set_option('display.max_columns', None)

In [5]:
test_query = "SELECT * FROM nyc_schools.high_school_directory LIMIT 5;"
pd.read_sql(test_query, engine)

dbn                                        school_name    borough  \
0  27Q260          Frederick Douglass Academy VI High School     Queens   
1  21K559        Life Academy High School for Film and Music   Brooklyn   
2  16K393     Frederick Douglass Academy IV Secondary School   Brooklyn   
3  08X305                               Pablo Neruda Academy      Bronx   
4  03M485  Fiorello H. LaGuardia High School of Music & A...  Manhattan   

  building_code  phone_number    fax_number  grade_span_min  grade_span_max  \
0          Q465  718-471-2154  718-471-2890             9.0              12   
1          K400  718-333-7750  718-333-7775             9.0              12   
2          K026  718-574-2820  718-574-2821             9.0              12   
3          X450  718-824-1682  718-824-1663             9.0              12   
4          M485  212-496-0700  212-724-5748             9.0              12   

  expgrade_span_min expgrade_span_max start_time end_time  \
0              None              None    7:45 AM  2:05 PM   
1              None              None    8:15 AM  3:00 PM   
2              None              None    8:00 AM  2:20 PM   
3              None              None    8:00 AM  3:50 PM   
4              None              None    8:00 AM  4:00 PM   

                                          priority01  \
0  Priority to Queens students or residents who a...   
1  Priority to New York City residents who attend...   
2                 Priority to continuing 8th graders   
3  Priority to Bronx students or residents who at...   
4                    Open to New York City residents   

                                          priority02  \
0  Then to New York City residents who attend an ...   
1                    Then to New York City residents   
2  Then to Brooklyn students or residents who att...   
3  Then to New York City residents who attend an ...   
4  Admission is based on the outcome of a competi...   

                                          priority03  \
0               Then to Queens students or residents   
1                                                      
2  Then to New York City residents who attend an ...   
3                Then to Bronx students or residents   
4  Students must audition for each program (studi...   

                                          priority04  \
0                    Then to New York City residents   
1                                                      
2             Then to Brooklyn students or residents   
3                    Then to New York City residents   
4  Students must be residents of New York City at...   

                        priority05 priority06 priority07 priority08  \
0                                                                     
1                                                                     
2  Then to New York City residents                                    
3                                                                     
4                                                                     

  priority09 priority10 location phone_number2 school_email  \
0                           None          None         None   
1                           None          None         None   
2                           None          None         None   
3                           None          None         None   
4                           None          None         None   

                                        website  \
0  http://schools.nyc.gov/schoolportals/27/Q260   
1  http://schools.nyc.gov/schoolportals/21/K559   
2  http://schools.nyc.gov/schoolportals/16/K393   
3                    www.pablonerudaacademy.org   
4                           www.laguardiahs.org   

                                              subway  \
0                       A to Beach 25th St-Wavecrest   
1  D to 25th Ave ; N to Ave U ; N to Gravesend - ...   
2             J to Kosciusko St ; M, Z to Myrtle Ave   
3                           

In [6]:
test_query2 = "SELECT * FROM nyc_schools.school_demographics LIMIT 5;"
pd.read_sql(test_query2, engine)

dbn                       Name  schoolyear fl_percent  frl_percent  \
0  01M015  P.S. 015 ROBERTO CLEMENTE    20052006       89.4          NaN   
1  01M015  P.S. 015 ROBERTO CLEMENTE    20062007       89.4          NaN   
2  01M015  P.S. 015 ROBERTO CLEMENTE    20072008       89.4          NaN   
3  01M015  P.S. 015 ROBERTO CLEMENTE    20082009       89.4          NaN   
4  01M015  P.S. 015 ROBERTO CLEMENTE    20092010                    96.5   

   total_enrollment prek   k grade1 grade2 grade3 grade4 grade5 grade6 grade7  \
0               281   15  36     40     33     38     52     29     38          
1               243   15  29     39     38     34     42     46                 
2               261   18  43     39     36     38     47     40                 
3               252   17  37     44     32     34     39     49                 
4               208   16  40     28     32     30     24     38                 

  grade8 grade9 grade10 grade11 grade12  ell_num  ell_percent  sped_num  \
0                                             36         12.8        57   
1                                             38         15.6        55   
2                                             52         19.9        60   
3                                             48         19.0        62   
4                                             40         19.2        46   

   sped_percent  ctt_num  selfcontained_num  asian_num  asian_per  black_num  \
0          20.3       25                  9         10        3.6         74   
1          22.6       19                 15         18        7.4         68   
2          23.0       20                 14         16        6.1         77   
3          24.6       21                 17         16        6.3         75   
4          22.1       14                 14         16        7.7         67   

   black_per  hispanic_num  hispanic_per  white_num  white_per  male_num  \
0       26.3           189          67.3          5        1.8       158   
1       28.0           153          63.0          4        1.6       140   
2       29.5           157          60.2          7        2.7       143   
3       29.8           149          59.1          7        2.8       149   
4       32.2           118          56.7          6        2.9       124   

   male_per  female_num  female_per  
0      56.2         123        43.8  
1      57.6         103        42.4  
2      54.8         118        45.2  
3      59.1         103        40.9  
4      59.6          84        40.4

In [7]:
test_query3 = "SELECT * FROM nyc_schools.school_safety_report LIMIT 5;"
pd.read_sql(test_query3, engine)

school_year building_code     dbn                 location_name  \
0     2015-16          K006  17K006                      P.S. 006   
1     2015-16          K005  16K005    P.S. 005 Dr. Ronald Mcnair   
2     2015-16          K003  13K003  P.S. 003 The Bedford Village   
3     2015-16          K002  75K141                     P.S. K141   
4     2015-16          K002  84K704    EXPLORE CHARTER SCHOOL(BS)   

  location_code              address borough  geographical_district_code  \
0          K006     43 SNYDER AVENUE       K                        17.0   
1          K005   820 HANCOCK STREET       K                        16.0   
2          K003  50 JEFFERSON AVENUE       K                        13.0   
3          K141  655 PARKSIDE AVENUE       K                        17.0   
4          K704  655 PARKSIDE AVENUE       K                        17.0   

  register                              building_name num_schools  \
0      712                                                   None   
1      231                                                   None   
2      450                                                   None   
3      374  655 PARKSIDE AVENUE CONSOLIDATED LOCATION        None   
4      529  655 PARKSIDE AVENUE CONSOLIDATED LOCATION        None   

                                 schools_in_building  major_n  oth_n  \
0                                           P.S. 006      0.0    1.0   
1                         P.S. 005 Dr. Ronald Mcnair      0.0    0.0   
2                       P.S. 003 The Bedford Village      0.0    0.0   
3  Parkside Preparatory Academy |  P.S. K141 |Exp...      NaN    NaN   
4  Parkside Preparatory Academy |  P.S. K141 |Exp...      NaN    NaN   

   nocrim_n  prop_n  vio_n engroupa   rangea  avgofmajor_n  avgofoth_n  \
0       2.0     1.0    0.0       4C  501-750          0.28        1.11   
1       2.0     0.0    0.0       2C    1-250          0.24        0.71   
2       2.0     0.0    0.0       3C  251-500          0.27        0.90   
3       NaN     NaN    NaN       3C  251-500           NaN         NaN   
4       NaN     NaN    NaN       4C  501-750           NaN         NaN   

   avgofnocrim_n  avgofprop_n  avgofvio_n borough_name  postcode   latitude  \
0           1.74         0.75        0.53    BROOKLYN    11226.0  40.648590   
1           1.71         0.51        0.32    BROOKLYN    11233.0  40.685695   
2           1.67         0.63        0.44    BROOKLYN    11216.0  40.682663   
3            NaN          NaN         NaN    BROOKLYN    11226.0  40.656183   
4            NaN          NaN         NaN    BROOKLYN    11226.0  40.656183   

   longitude  community_board  council_district  census_tract        bin  \
0 -73.957138             14.0              40.0         794.0  3117230.0   
1 -73.922463              3.0              41.0         377.0  3039906.0   
2 -73.955338              3.0              36.0         227.0  3057406.0   
3 -73.951583              9.0              40.0         802.0  3378201.0   
4 -73.951583              9.0              40.0         802.0  3378201.0   

            bbl                                                nta  _schools  
0  3.051030e+09  Erasmus                                       ...         1  
1  3.014900e+09  Stuyvesant Heights                            ...         1  
2  3.019980e+09  Clinton Hill                                  ...         1  
3  3.050500e+09  Prospect Lefferts Gardens-Wingate             ...         3  
4  3.050500e+09  Prospect Lefferts Gardens-Wingate             ...         3

#🧮 School Distribution

How many schools are there in each borough?

In [8]:
query_school_count = """
SELECT borough,
       COUNT(DISTINCT school_name) AS unique_school_count
FROM nyc_schools.high_school_directory
WHERE borough IS NOT NULL
GROUP BY borough
ORDER BY unique_school_count DESC, borough;
"""
df_school_count = pd.read_sql(query_school_count, engine)
print(df_school_count)
#this query extracts borough and the count of unique school names from the high_school_directory grouped by borough
#the schools get counted accordingly to the borough in the row resulting in a total N

         borough  unique_school_count
0       Brooklyn                  121
1          Bronx                  118
2      Manhattan                  106
3         Queens                   80
4  Staten Island                   10


In [11]:
query_school_count = """
SELECT
  h.borough,
  COUNT(DISTINCT h.dbn) AS n_schools
FROM nyc_schools.high_school_directory AS h
WHERE h.borough IS NOT NULL
GROUP BY h.borough
ORDER BY n_schools DESC, h.borough;
"""
df_school_count = pd.read_sql(query_school_count, engine)
print(df_school_count)

         borough  n_schools
0       Brooklyn        121
1          Bronx        118
2      Manhattan        106
3         Queens         80
4  Staten Island         10


Most schools are located in Brooklyn(121) closely followed by Bronx(118). Staten Island has only 10.

#🎓 Language Learners

What is the average % of English Language Learners (ELL) per borough?

In [9]:
query_ell_avg_borough = """
SELECT h.borough,
       ROUND(AVG(d.ell_percent)::numeric, 2) AS avg_ell_percent
FROM nyc_schools.school_demographics d
JOIN nyc_schools.high_school_directory h
  ON d.dbn = h.dbn
WHERE d.ell_percent IS NOT NULL
GROUP BY h.borough
ORDER BY h.borough;
"""
df_ell = pd.read_sql(query_ell_avg_borough, engine)
print(df_ell)
#here borough and the average ell_percent get extracted from two tables
#the tables get joined on dbn to match entrys

     borough  avg_ell_percent
0  Manhattan             7.57


Only Manhattan has data for english learning students!

#🔗School supporting special needs

Using the data from the school demographics and high school directory, write a query to find the top 3 schools in each borough with the highest percentage of special education students (sped_percent)

In [10]:
query_top_sped = """
SELECT dir.borough,
       dir.school_name,
       MAX(demo.sped_percent) AS sped_percent
FROM nyc_schools.high_school_directory dir
JOIN nyc_schools.school_demographics demo
  ON dir.dbn = demo.dbn
WHERE demo.sped_percent IS NOT NULL
GROUP BY dir.borough, dir.school_name
ORDER BY MAX(demo.sped_percent) DESC
LIMIT 3;
"""

df_top_sped = pd.read_sql(query_top_sped, engine)
print(df_top_sped)

#joining directory and demographics table to get the top 3 schools with highest percentage of speced students

     borough                                    school_name  sped_percent
0  Manhattan                     East Side Community School          28.8
1  Manhattan                        Marta Valle High School          25.9
2  Manhattan  Henry Street School for International Studies          25.1


In [16]:
query_top3_sped_per_borough = """
WITH ranked AS (
  SELECT
    h.borough,
    h.school_name,
    h.dbn,
    d.sped_percent,
    ROW_NUMBER() OVER (
      PARTITION BY h.borough
      ORDER BY d.sped_percent DESC, h.school_name
    ) AS rn
  FROM nyc_schools.school_demographics d
  JOIN nyc_schools.high_school_directory h
    ON d.dbn = h.dbn
  WHERE d.sped_percent IS NOT NULL
    AND h.borough IS NOT NULL
)
SELECT borough, school_name, dbn, sped_percent
FROM ranked
WHERE rn <= 3
ORDER BY borough, rn;
"""

df_top_sped = pd.read_sql(query_top3_sped_per_borough, engine)
print(df_top_sped)


     borough                 school_name     dbn  sped_percent
0  Manhattan  East Side Community School  01M450          28.8
1  Manhattan  East Side Community School  01M450          27.7
2  Manhattan  East Side Community School  01M450          26.7


Only Manhattan has Data available for speced students!